catalog/transformer/catalog-transformer-csv-queryresponse/src/main/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformer.java

```diff
@@ -0,0 +1,111 @@
+import java.util.stream.Collectors;
+
+/**
+ * An implementation of QueryResponseTransformer that produces CSV output.
+ *
+ * @see ddf.catalog.transform.QueryResponseTransformer
+ */
+public class CsvQueryResponseTransformer implements QueryResponseTransformer {
+  private static final String HIDDEN_FIELDS_KEY = "hiddenFields";
+
+  private static final String COLUMN_ORDER_KEY = "columnOrder";
+
+  private static final String COLUMN_ALIAS_KEY = "aliases";
+
+  /**
+   * @param upstreamResponse the SourceResponse to be converted.
+   * @param arguments this transformer accepts 3 parameters in the 'arguments' map. 1) key:
+   *     'hiddenFields' value: a java.util.Set containing Attribute names (as Strings) to be
+   *     excluded from the output. 2) key: 'attributeOrder' value: a java.utilList containing
+   *     Attribute name (as Strings) to identify the order that the columns will appear in the
+   *     output. 3) key: 'aliases' value: a java.util.Map whose keys are attribute names and values
+   *     are the how that attribute column should be aliased in the output. For example, if the key
+   *     is 'title' and the value is 'Product' then the resulting CSV will have a column name of
+   *     'Product' instead of 'title'.
+   * @return a BinaryContent object that contains an InputStream with the CSV content.
+   * @throws CatalogTransformerException during processing, the CSV output is written to an
+   *     Appendable, whose 'append()' method signature declares that it throws IOException. When
+   *     that Appendable throws IOException, this class will theoretically convert that into a
+   *     CatalogTransformerException and raise that. Because this implementation uses a
+   *     StringBuilder which doesn't throw IOException, this will never occur.
+   */
+  @Override
+  public BinaryContent transform(
+      SourceResponse upstreamResponse, Map<String, Serializable> arguments)
+      throws CatalogTransformerException {
+
+    List<Metacard> metacards =
+        upstreamResponse
+            .getResults()
+            .stream()
+            .map(Result::getMetacard)
+            .collect(Collectors.toList());
+
+    Set<String> hiddenFields =
+        Optional.ofNullable((Set<String>) arguments.get(HIDDEN_FIELDS_KEY))
+            .orElse(Collections.emptySet());
+
+    List<String> attributeOrder =
+        Optional.ofNullable((List<String>) arguments.get(COLUMN_ORDER_KEY))
+            .orElse(Collections.emptyList());
+
+    Map<String, String> columnAliasMap =
+        Optional.ofNullable((Map<String, String>) arguments.get(COLUMN_ALIAS_KEY))
+            .orElse(Collections.emptyMap());
+
+    Set<AttributeDescriptor> allAttributeDescriptors = getAllAttributes(metacards, hiddenFields);
+
+    Set<String> requestedFields = new HashSet<>(attributeOrder);
+
+    // If requestedFields is not empty, additionally filter out all non-requested attributes.
+    if (!requestedFields.isEmpty()) {
+      allAttributeDescriptors =
```

Comment:
✏️ I think `allAttributeDescriptors.removeIf(ad -> !requestedFields.contains(ad.getName()));` would also work.

Reply:
That public static method `CsvTransformer.getOnlyRequestedAttributes` might also be useful here.